### PR TITLE
Initial Work on ZIP155 Backport of BIP155

### PR DESCRIPTION
**Pull request by zancas**
As mentioned in https://github.com/zcash/zcash/pull/5276 we're backporting the minimum lines of networking code we can in order to support TorV3 addresses.

From the perspective of the bitcoin developers the process of implementing BIP155 (a prereq for TorV3 addresses) began with the sanity check here:  https://github.com/bitcoin/bitcoin/pull/19351.

Our task began with the minimal reconciliation of zcash with bitcoin such that this sanity check could be used.

The following is some of that work.   Our hope is that reviewing this backported code will be relatively simple, such that it can be easily/quickly incorporated into zcash, if we're on the right track.  If, on the other hand, there's a significant problem with this approach, we'd like to know ASAP.

This PR is derived from the following upstream PRs, listed in chronological order:
 1. https://github.com/bitcoin/bitcoin/pull/7257:
      -- small change adding logging helper functions
 2. https://github.com/bitcoin/bitcoin/pull/7868:
      -- beginning of network code refactor, removes name resolution from storage classes
 3. https://github.com/bitcoin/bitcoin/pull/7749/commits/ee06e04369c37da21e048fda849cce2a1f066f84:
      -- single commit from 7749 adding a ServiceFlags enum definition in place of raw uint64_t representation of Service offerings

### CODE REVIEW

File: src/addrman.cpp [hunk 1]
```diff
@@ -259,7 +259,7 @@ bool CAddrMan::Add_(const CAddress& addr, const CNetAddr& source, int64_t nTimeP
             pinfo->nTime = std::max((int64_t)0, addr.nTime - nTimePenalty);
 
         // add services
-        pinfo->nServices |= addr.nServices;
+        pinfo->nServices = ServiceFlags(pinfo->nServices | addr.nServices);
```
Comment (zancas) [inline src/addrman.cpp:262]: `nServices` (a `CAddress` member) is now a `ServiceFlags`.

File: src/addrman.h [hunk 2]
```diff
@@ -569,12 +569,13 @@ class CAddrMan
...
-        LOCK(cs);
-        Check();
-        Connected_(addr, nTime);
-        Check();
+        {
+            LOCK(cs);
+            Check();
+            Connected_(addr, nTime);
+            Check();
+        }
```
Comment (zancas) [inline src/addrman.h:577]: Presumably this is to do with dropping the lock?  It seems redundant, but I suppose one might later add code outside of the inner block, but still in the function.

File: src/init.cpp [hunk 3]
```diff
@@ -837,6 +837,16 @@ void InitParameterInteraction()
...
+static std::string ResolveErrMsg(const char * const optname, const std::string& strBind)
+{
+    return strprintf(_("Cannot resolve -%s address: '%s'"), optname, strBind);
+}
+
+static std::string AmountErrMsg(const char * const optname, const std::string& strValue)
+{
+    return strprintf(_("Invalid amount for -%s=<amount>: '%s'"), optname, strValue);
+}
+
```
Comment (zancas) [inline src/init.cpp:849]: These helpers and their calls simply redefine the interface to a few Logged messages.

File: src/init.cpp [hunk 4]
```diff
@@ -1419,11 +1429,12 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
     }
 
     if (mapArgs.count("-externalip")) {
-        for (const std::string& strAddr : mapMultiArgs["-externalip"]) {
-            CService addrLocal(strAddr, GetListenPort(), fNameLookup);
-            if (!addrLocal.IsValid())
-                return InitError(strprintf(_("Cannot resolve -externalip address: '%s'"), strAddr));
-            AddLocal(CService(strAddr, GetListenPort(), fNameLookup), LOCAL_MANUAL);
+        BOOST_FOREACH(const std::string& strAddr, mapMultiArgs["-externalip"]) {
```
Comment (zancas) [inline src/init.cpp:1432]: We choose to prefer `BOOST_FOREACH` because that seems to be closer to bitcoin.

File: src/net.cpp [hunk 5]
```diff
@@ -1409,18 +1409,26 @@ void ThreadDNSAddressSeed()
...
-            addrman.Add(vAdd, CNetAddr(seed.name, true));
+            // TODO: The seed name resolve may fail, yielding an IP of [::], which results in
+            // addrman assigning the same source to results from different seeds.
+            // This should switch to a hard-coded stable dummy IP for each seed name, so that the
+            // resolve is not required at all.
+            if (!vIPs.empty()) {
+                CService seedSource;
+                Lookup(seed.name.c_str(), seedSource, 0, true);
+                addrman.Add(vAdd, seedSource);
+            }
```
Comment (zancas) [inline src/net.cpp:1431]: The check for the existence of `vIPs` used to occur _within_ (inside `Lookup`) the constructor call stack for `CNetAddr`, it has been factored out.
This means it's the caller's responsibility.

File: src/net.cpp [hunk 6]
```diff
@@ -1636,7 +1644,9 @@ void ThreadOpenAddedConnections()
         for (vector<CService>& vserv : lservAddressesToAdd)
         {
             CSemaphoreGrant grant(*semOutbound);
-            OpenNetworkConnection(CAddress(vserv[i % vserv.size()]), &grant);
+            /* We want -addnode to work even for nodes that don't provide all
+             * wanted services, so pass in nServices=NODE_NONE to CAddress. */
+            OpenNetworkConnection(CAddress(vserv[i % vserv.size()], NODE_NONE), &grant);
```
Comment (zancas) [inline src/net.cpp:1649]: We preferred zcash's version of `OpenNetworkConnection` to minimize the size of the backport (thus far).

File: src/netbase.cpp [hunk 7]
```diff
@@ -615,10 +615,12 @@ bool ConnectSocketByName(CService &addr, SOCKET& hSocketRet, const char *pszDest
     proxyType nameProxy;
     GetNameProxy(nameProxy);
 
-    CService addrResolved(CNetAddr(strDest, fNameLookup && !HaveNameProxy()), port);
-    if (addrResolved.IsValid()) {
-        addr = addrResolved;
-        return ConnectSocket(addr, hSocketRet, nTimeout);
+    CService addrResolved;
```
Comment (zancas) [inline src/netbase.cpp:618]: `Lookup` is no longer part of construction.

File: src/netbase.cpp [hunk 8]
```diff
@@ -685,19 +687,19 @@ CNetAddr::CNetAddr(const struct in6_addr& ipv6Addr)
     SetRaw(NET_IPV6, (const uint8_t*)&ipv6Addr);
 }
 
-CNetAddr::CNetAddr(const char *pszIp, bool fAllowLookup)
+CNetAddr::CNetAddr(const char *pszIp)
```
Comment (zancas) [inline src/netbase.cpp:690]: `Lookup` is no longer a constructor concern.

File: src/netbase.h [hunk 9]
```diff
@@ -48,8 +48,8 @@ class CNetAddr
     public:
         CNetAddr();
         CNetAddr(const struct in_addr& ipv4Addr);
-        explicit CNetAddr(const char *pszIp, bool fAllowLookup = false);
-        explicit CNetAddr(const std::string &strIp, bool fAllowLookup = false);
+        explicit CNetAddr(const char *pszIp);
```
Comment (zancas) [inline src/netbase.h:51]: `Lookup` is no longer a constructor concern.

File: src/protocol.h [hunk 10]
```diff
@@ -104,13 +106,15 @@ class CAddress : public CService
         if ((s.GetType() & SER_DISK) ||
             (nVersion >= CADDR_TIME_VERSION && !(s.GetType() & SER_GETHASH)))
             READWRITE(nTime);
-        READWRITE(nServices);
+        uint64_t nServicesInt = nServices;
+        READWRITE(nServicesInt);
+        nServices = (ServiceFlags)nServicesInt;
```
Comment (zancas) [inline src/protocol.h:111]: Hack to work around the `READWRITE` interface (I guess).

File: src/addrman.h (hunk 2 above)
Comment (AloeareV) [reply src/addrman.h:577]: By 'this' you mean the double braces? I've seen them around a lot, where there's a block that contains nothing except for another block. As far as I can tell it's semantically meaningless, but I could be missing something, or they're some sort of human-meaningful convention I've never heard of.

**Comment by teor2345** [comment]
From quickly skimming this PR, it looks like it doesn't change `zcashd`'s network behaviour at all.
(Except perhaps for changing the timing of DNS resolution.)

Would you mind updating the PR description to summarise the changes in the PR?
(Usually 1 sentence is enough.)

If you can do that for future PRs, then the Zebra developers will know which PRs we can test our `addrv2` implementation against. (Although you might be finished and merged before we start implementing!)

**Comment by str4d** [comment]
Now that v4.5.1 is out of the way, I'll be adding this to my review stack.

**Comment by zancas** [comment]
I'm reorganizing the relevant code, I will open a new PR to track the reorg.

**Pull request closed without merging**